docs/enhancements/fep-002-config-hooks.md

```diff
@@ -0,0 +1,204 @@
+ "config_name": "my-config",
+ "device_names": [`device-name1`, `device-name2`],
+]
+
+
+### The hook events
+
+#### FlightCTL to external system
+* `ON_FLEET_JOIN`: This event is triggered when a device is added to a fleet for
+  the first time. The external system can use this event to provide the initial
+  configuration or to start building any specific configuration that needs to be
+  created.
+
+* `ON_DEVICE_SPEC_UPDATE`: This event is triggered when a device.spec is updated. At this
+  point, if the device spec contains an ExternalConfigProviderSpec, the external system
+  can be called to get the new configuration.
+
+* `ON_DEVICE_METADATA_UPDATE`: This event is triggered when a device.metadata is updated. At this
+  point, if the device spec contains an ExternalConfigProviderSpec, the external system
+  can be called to let it know about metadata changes: labels, etc.
+
+* `ON_DEVICE_POLL`: This event is triggered by flightctl device-controller periodically
+  to check if the device configuration injected from this external system is still valid
+  the periodicty is decided by the external system on previous hook calls.
+
+* `ON_ENROLLMENT_REQUEST`: This is an event that is triggered when a device waiting
+  for enrollment request is targeted to this fleet because of the labels being
+  proposed on the device. This is an early notification that can be used to start
+  constructing credentials or triggering any other action.
+
+#### External system to FlightCTL
+When calling the external system flightctl provides a callback endpoint, as well as the
+necessary authorization (like a token) to call back to flightctl. The external system
+can notify flightctl about the following events:
+
+* `ON_CONFIG_UPDATE`: This event is triggered when the external system wants an specific
+ device configuration to be updated: i.e. new credentials for the network, a token
+ being removed because the application has finished onboarding, etc. This event
+ can point to a single device, or to a list of devices.
+
+
+
+
+### User Stories
+
+#### Story 1
+As a Fleet administrator, I want my Microshift based devices on a specific fleet
+to automatically inject klusterlet details, connection tokens or any credentials
+automatically, just by pointing my fleet to a service running on the ACM cluster.
+
+#### Story 2
+As a IT user, I want to have a central webhook that distributes network credentials
+to my devices, and has the logic to know which credentials belong to which device
+based on device attributes (location, serial number, region..). While this could
+be handled with templating, we need additional flexibility or templating cannot
+
+capture our needs well enough.
+
+### Risks and Mitigations
+
+Fleets consuming this introduce an external dependency that needs to be called for every
+device. This could be mitigated by having a local cache of the configuration on the device
+and only calling the external system when the cache is stale.
```

Comment:
We should probably log and emit metrics related to those types of endpoints, so the admin can understand response times and any potential slow-down related to those services.